### PR TITLE
upgrade workflow

### LINKED ISSUE #2171: Basic upgrade workflow
Here is the upgrade workflow
- Find existing VCH appliance
- Query out VCH configuration
- Verify new version and existing VCH's version
- Valiate VCH configuration with new version's validator. If anything incompatible found, stop the upgrade. As no data migration yet.
- Snapshot VCH appliance VM. - Should quiesce VCH during upgrade here (how?)
- Poweroff VCH appliance VM if it's not
- Upload iso files
- Generate new VCH configuration from old one
- Update VCH configuration version
- Update container bootstrap iso file path
- Update container VM configuration for anything required to support VCH restart(e.g. vch ip change, vch serial port number change, etc.)
- Attach new appliance iso to VCH appliance VM
- Power on VCH, and wait VCH initialization
- Anything wrong in the above steps, switch appliance iso file back to old one, and then roll back to upgrade snapshot

### PR DESCRIPTION
**Pull request by emlin**
Fixes #2171 

create snapshot, upload new iso files, switch appliance VM iso files, and then update configuration. If failed, rollback to snapshot, and clean up everything created.

upgrade is not finished in this change. So upgrade command will not be shown in command help

### CODE REVIEW

File: cmd/vic-machine/inspect/inspect.go [hunk 1]
```diff
@@ -28,7 +28,7 @@ import (
 	"golang.org/x/net/context"
 )
 
-// Delete has all input parameters for vic-machine delete command
+// Delete has all input parameters for vic-machine inspect command
```
Comment (andrewtchin) [inline cmd/vic-machine/inspect/inspect.go]: `// Inspect has all input parameters for vic-machine inspect command`

File: lib/install/management/upgrade.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+	snapshotRefID, err := d.createSnapshot(snapshotName, "upgrade snapshot")
+	if err != nil {
+		return err
+	}
+
+	if err = d.update(conf, settings); err == nil {
+		return nil
+	}
+	log.Errorf("Failed to upgrade: %s", err)
+	log.Infof("Roll back upgrade")
```
Comment (andrewtchin) [inline lib/install/management/upgrade.go]: `Rolling back upgrade`

File: lib/install/management/upgrade.go [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+	return err
+}
+
+func (d *Dispatcher) cleanSnapshot(id types.ManagedObjectReference, name string) error {
+	log.Infof("Deleting upgrade snapshot %q", name)
+	if _, err := tasks.WaitForResult(d.ctx, func(ctx context.Context) (tasks.Task, error) {
+		return d.appliance.RemoveSnapshot(ctx, id, true, true)
+	}); err != nil {
+		log.Errorf("Failed to clean up appliance upgrade snapshot %q: %s.", name, err)
+		log.Errorf("Be sure to remove snapshot %q manually before upgrade again", name)
```
Comment (andrewtchin) [inline lib/install/management/upgrade.go]: ```log.Errorf("Snapshot %q MUST be removed manually before attempting upgrade again", name)```

Comment (andrewtchin) [inline lib/install/management/upgrade.go]: If it's not too hard maybe include the name of the VM so the user can navigate to the snapshot

**Comment by andrewtchin** [comment]
lgtm

File: cmd/vic-machine/common/images.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+	}
+	imgs[ApplianceImageName] = i.ApplianceISO
+
+	if i.BootstrapISO == "" {
+		i.BootstrapISO = osImgs[0]
+	}
+	imgs[LinuxImageName] = i.BootstrapISO
+
+	for name, img := range imgs {
+		_, err := os.Open(img)
```
Comment (jzt) [inline cmd/vic-machine/common/images.go]: Can you not use `os.Stat` here, too? Does the unused filehandle get `Close()`d in this case?

**Comment by jzt** [comment]
Other than my minor comment, LGTM

File: cmd/vic-machine/common/images.go (hunk 4 above)
Comment (emlin) [inline cmd/vic-machine/common/images.go]: good point

**Comment by emlin** [comment]
♻️

**Pull request merged**